### DISCUSSION TITLE
Parse multipart form data - config syntax

### DISCUSSION THREAD

**Discussion by LumpiStefan** [adnanh/webhook#537] [Q&A]
I´m new in the webhook topic and I try to implement my first (or to be more precise my second) webhook.
The webhook i receieve is a multipart form:
```
Content-Type: multipart/form-data; boundary=44a89c4cb58f9fb56ad712672330d4d5
```
First part is a file (e.g. PNG):
```
Content-Disposition: form-data; name="screenshot"; filename="Screenshot.png"
Content-Type: image/png
<DATA>
```
Second Part is a json:
```
Content-Disposition: form-data; name="feedback"
Content-Type: application/json
{"feedback_number": 433, "link": "https://.....
```

Now I have several problems/questions:
1. How can I save the file that is included in the webhook?
2. How can I pass the information of the json to the execute-command?
What is the correct syntax for the hook-configuration-file?

I tried:
```
...
"parse-parameters-as-json": [
      {
        "source": "payload",
        "name": "feedback"
      }
    ],
    "pass-arguments-to-command":
      [
        {
           "source": "feedback",
           "name": "feedback.feedback_number"
        },
...
```
but it is not working.
What has to be used for the source and the name fields?

Sorry if that are stupid questions, but in the moment I´m little bit lost...

Regards
Stefan

**Comment by moorereason** [comment]
To parse the "feedback" part as JSON, I think you would use:

```
"parse-parameters-as-json": [
  {
    "source": "payload",
    "name": "feedback"
  }
],
```

See https://github.com/adnanh/webhook/blob/master/docs/Hook-Examples.md#multipart-form-data

Saving the "screenshot" as a PNG may be trickier, but you should be able to use `pass-file-to-command` for that.  I'd have to play with it which I don't have time to do right now.

See https://github.com/adnanh/webhook/blob/master/docs/Hook-Examples.md#pass-file-to-command-sample

**Comment by LumpiStefan** [comment]
Thanks for the answer. Sadly it does not work.
Currently my setting is:
```
[
  {
    "id": "usersnap",
    "execute-command": "/var/scripts/usersnap.sh",
    "command-working-directory": "/var/webhook",
    "parse-parameters-as-json": [
      {
        "source": "payload",
        "name": "feedback"
      }
    ],
    "pass-arguments-to-command":
      [
        {
           "source": "feedback",
           "name": "feedback.feedback_number"
        },
        {
           "source": "feedback",
           "name": "feedback_number"
        },
```

Output von webhook:
```
[webhook] 2021/06/15 08:20:13 [2c12a3] incoming HTTP POST request from 18.196.29.249:35874
[webhook] 2021/06/15 08:20:13 [2c12a3] usersnap got matched
[webhook] 2021/06/15 08:20:14 [2c12a3] found multipart form value "feedback"
[webhook] 2021/06/15 08:20:14 [2c12a3] usersnap hook triggered successfully
[webhook] 2021/06/15 08:20:14 [2c12a3] 200 | 0 B | 65.591615ms | 85.25.44.68:8888 | POST /hooks/usersnap
[webhook] 2021/06/15 08:20:14 [2c12a3] error extracting command arguments: couldn't retrieve argument for {Source:feedback Name:feedback.feedback_number EnvName: Base64Decode:false}
[webhook] 2021/06/15 08:20:14 [2c12a3] error extracting command arguments: couldn't retrieve argument for {Source:feedback Name:feedback_number EnvName: Base64Decode:false}
```

In the given example (https://github.com/adnanh/webhook/blob/master/docs/Hook-Examples.md#multipart-form-data) there is often the reference to "payload" as source and name. For me the handling is unclear.
Maybe it would be more clear whe there is described how the source and names are handeled...

Stefan

**Reply by moorereason** [reply]
Try using `"source": "payload"`

**Reply by LumpiStefan** [reply]
Ok, that looks now better. Thanks for the help so long...
` "parse-parameters-as-json": [
      {
        "source": "payload",
        "name": "feedback"
      }
    ],
    "pass-arguments-to-command":
      [
        {
           "source": "payload",
           "name": "feedback.feedback_numer"
        },
`
is working fine. Now my last problem is extracting the attached file.
Therefore i have configured:
`"pass-file-to-command": [
      {
        "source": "payload",
        "name": "screenshot",
        "filename": "binaryFile.zip",
        "base64decode": true
      }
      ]`
Content of the webhook was:
`Content-Disposition: form-data; name="screenshot"; filename="Screenshot.png"
Content-Type: image/png`

Output of webhook:
[webhook] 2021/06/17 18:40:47 [b3da37] error extracting command arguments for file: couldn't retrieve argument for {Source:payload Name:screenshot EnvName: Base64Decode:true}

Any idea on this case?

Regards
Stefan
